cloudferrylib/os/network/neutron.py

```diff
@@ -1196,3 +1197,67 @@ def get_new_extnet_id(self, src_net_id, src_nets, dst_nets):
             net_hash = self.get_res_hash_by_id(src_nets, src_net_id)
             dst_net_id = self.get_res_by_hash(dst_nets, net_hash)['id']
         return dst_net_id
+
+    @staticmethod
+    def get_free_ip(subnet, excluded_ips):
+        for pool in subnet['allocation_pools']:
+            addr = ipaddr.IPAddress(pool['start'])
+            end = ipaddr.IPAddress(pool['end'])
+            while addr <= end:
+                if str(addr) not in excluded_ips:
+                    return str(addr)
+                addr += 1
+        raise Exception("Can't find free ip for router")
+
+    @staticmethod
+    def get_routers_for_changing(info):
+        result = []
+        if "routers" in info:
+            subnets = {subnet['id']: subnet for subnet in info['subnets']}
+            allocated_ip = collections.defaultdict(list)
+            if 'floating_ips' in info:
+                for floating_ip in info['floating_ips']:
+                    allocated_ip[floating_ip['floating_network_id']].append(
+                        floating_ip['floating_ip_address'])
+            routers = []
+            for router_info in info['routers']:
+                router = Router(router_info, subnets)
+                routers.append(router)
+                allocated_ip[router.ext_net_id].append(router.ext_ip)
+            for router in routers:
+                router.new_ip = NeutronNetwork.get_free_ip(
+                    subnets[router.ext_subnet_id],
+                    allocated_ip[router.ext_net_id])
+                allocated_ip[router.ext_net_id].append(router.new_ip)
+                result.append(router.to_map())
+        return result
+
+    def get_router_host(self, router_id):
```

Comment:
One router can be placed only on one controller. Otherwise we will have two routers with the same IP in one cloud.
Here I take network controller for each router. It's cover case when we have more controllers and routers distributed to them.